lib/puppet/pops/types/enumeration.rb

```diff
@@ -1,33 +1,72 @@
 # The Enumeration class provides default Enumerable::Enumerator creation for Puppet Programming Language
 # runtime objects that supports the concept of enumeration.
 #
-class Puppet::Pops::Types::Enumeration
-  # Produces an Enumerable::Enumerator for Array, Hash, Integer, Integer Range, and String.
+module Puppet::Pops::Types
+class Enumeration
+
+  # Produces an `Enumerator` for `String`, `Integer`, `PIntegerType` or anything that respond to `each` by
+  # returning an `Enumerator`. The argument is returned in case it already is an `Enumerator`.
+  #
+  # @param caller [Function] The function that needs to coerce its argument to an enumeration
+  # @param o [Object] The object to produce an `Enumerator` for
+  # @return [Enumerator] The produced `Enumerator` or nil if it couldn't be produced
+  # @raise [ArgumentError] In case an `Enumerator` cannot be produced
+  # @api public
+  def self.asserted_enumerable(caller, o)
+    enum = self.enumerator(o)
+    raise ArgumentError, "#{caller.class.name}(): wrong argument type (#{o.class.name}; must be something enumerable." if enum.nil?
+    enum
+  end
+
+  # Produces an `Enumerator` for `String`, `Integer`, `PIntegerType` or anything that respond to `each` by
+  # returning an `Enumerator`. The argument is returned in case it already is an `Enumerator`.
   #
+  # @param o [Object] The object to produce an `Enumerator` for
+  # @return [Enumerator,nil] The produced `Enumerator` or `nil` if it couldn't be produced
+  # @api public
   def self.enumerator(o)
     @@singleton ||= new
     @@singleton.enumerator(o)
   end
 
-  # Produces an Enumerator for Array, Hash, Integer, Integer Range, and String.
+  # Produces an `Enumerator` for `String`, `Integer`, `Range`, `PIntegerType` or anything that respond to `each` by
+  # returning an `Enumerator`. The argument is returned in case it already is an `Enumerator` or an `Iterable`.
   #
+  # @param o [Object] The object to produce an `Enumerator` for
+  # @return [Enumerator,nil] The produced `Enumerator` or `nil` if it couldn't be produced
+  # @api public
   def enumerator(o)
     case o
```

Comment:
It would be nice to add iteration of all PTypes that contain types (but that is a separate ticket).